packages/svelte/src/internal/server/index.js

```diff
@@ -155,6 +155,22 @@ export function head(payload, fn) {
 	head_payload.out += BLOCK_CLOSE;
 }
 
+/**
+ * Create a snippet imperatively using mount, hyrdate and render functions.
+ * @param {{
+ * 	 mount: (...params: any[]) => Element,
+ *   hydrate?: (element: Element, ...params: any[]) => void,
+ *   render: (...params: any[]) => string
+ * }} options
+ */
+export function createRawSnippet({ render }) {
+	const snippet_fn = (/** @type {Payload} */ payload, /** @type {any[]} */ ...args) => {
+		payload.out += render(...args);
```

Comment:
Good idea.

Reply:
I'm not sure about this. You can't have `<svelte:head>` inside a `{#snippet ...}`, which is the equivalent, surely?

Reply:
@Rich-Harris You can have `<SomeComponent />` inside the snippet which has a `<svelte:head>` though?

Reply:
Actually, scrap that, Astro and others would likely just use `mount` or `hydrate`, right?

Reply:
They do, but on the server the use render: https://github.com/withastro/astro/blob/main/packages/integrations/svelte/server-v5.js#L32

Also, if you're using `mount` and `hydrate`, you get support for adding stuff to the head. So why wouldn't you be able to manually do that here aswell?

```js
	const snippet = createRawSnippet({
		mount(text) {
			const title = document.createElement('title');

			$effect(() => {
				title.textContent = text();
			});

			return document.createComment('');
		},
		hydrate(element, count) {
			$effect(() => {
				document.head.querySelector('title[data-x]').textContent = text();
			});
		},
		render(count) {
			return { html: '<!---->', head: '<title data-x>text</title>' }
		}
	});
```

(which then begs the question: How do you cleanup head content when the snippet is destroyed?)

Astro is currently not making use of properly having support for head content, but in theory they could add it I think? @bluwy is this in general possible in Astro for a framework to contribute to contents in the `<head>` tag?

Reply:
> Also, if you're using mount and hydrate, you get support for adding stuff to the head.

I mean I removed it because it complicated this and Astro wasn't making use of it, plus after implementing it I ran into countless hydration issues, maybe related to https://github.com/sveltejs/svelte/issues/12399.

I'm also a bit unsure how someone might hydrate the head in their `hydrate` function being able to control the internal node logic.

Reply:
> How do you cleanup head content when the snippet is destroyed?

With an effect, presumably?

I'm in two minds about this whole thing. It does feel like a limitation (regardless of whether Astro specifically would be able to take advantage of it), but getting it to work well would add a ton of complexity when you factor in hydration and stuff. Like, a simple thing like this...

```svelte
<p>before</p>
{#if condition}
  <div class="a"><A /></div>
{:else}
  <div class="b"><B /></div>
{/if}
<p>after</p>
```

...would presumably have to become this, if it's possible that `A` or `B` contains head content (directly or indirectly):

```js
render: (condition) => {
  let head = '';
  let body = '<p>before</p>';

  if (condition) {
    const rendered = render(A);
    body += '<div class="a">${rendered.body}</div>`
    head += rendered.head;
  } else {
    const rendered = render(B);
    body += '<div class="b">${rendered.body}</div>`
    head += rendered.head;
  }

  body += '<p>after</p>';
  
  return { head, body };
}
```

As opposed to this, if you just allow the constraint that programmatic snippets can't create head content:

```js
render: (condition) => {
  return (
    '<p>before</p>' +
    condition ? `<div class="a">${render(A).body}</div>` : `<div class="b">${render(B).body}</div>` +
    '<p>after</p>
  );
}
```

So the increased complexity disadvantages users more than us. I suppose the alternative is that just expose the `payload` object...

```js
render: (payload, condition) => {
  payload.out += '<p>before</p>';
  if (condition) {
    payload.out += '<div class="a">';
    A(payload);
    payload.out += '</div>';
  } else {
    payload.out += '<div class="b">';
    B(payload);
    payload.out += '</div>';
  }
  payload.out += '<p>after</p>';
}
```

...but that hardly feels like an improvement. So I think I land on where we are now — keeping it simple.

Reply:
> @bluwy is this in general possible in Astro for a framework to contribute to contents in the `<head>` tag?

At the moment I don't think it's possible, especially with regards to streaming, it's not easy to aggregate them beforehand unless Astro do something to detect if components exposes the head, which is tricky. I think it's ok to punt this off for now though.